### PR TITLE
implementing geocoding

### PR DESCRIPTION
**Pull request by pokyah**
Hi @koenderks ,

I have implemented a geocoding function that would allow to pass any location name.

Minimal example : 

```{r}
name = "Namur"
country = "Belgium"
theme = "original"
city = geocode(location = paste0(name, " ", country))

city <- new_city(
  name = name,
  country = country,
  lat = city[1, 3],
  long = city[1, 4]
)

map = cityview(
  name = city,
  license = F,
  theme = theme,
  timeout = 1000,
  legend = F,
  zoom = 3.75,
  border = "circle",
  halftone = "#ffffff",cache = T
)
```

Let me know if something is unclear or if I should implement it in another manner.

Thanks

### CODE REVIEW

File: R/geocode.R [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  if (method %in% methods_with_keys) {
+    api_key_env <- paste0(toupper(method), "_API_KEY")
+    api_key <- Sys.getenv(api_key_env)
+    if (api_key == "") {
+      stop(paste0("API key for ", method, " is required. Please set the '", api_key_env, "' environment variable."))
+    }
+  }
+  
+  result <- tryCatch({
+    geocode_df <- tibble::tibble(address = location) %>%
```
Comment (koenderks) [inline R/geocode.R]: `tidygeocoder::geocode(.tbl = data.frame(address=location), "address", method = method, quiet = TRUE)` gets rid of both the dependency on `tibble` and `dplyr`, which is nice.

File: R/geocode.R [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+#' @export
+geocode <- function(location, method = "osm") {
+  if (missing(location) || !is.character(location)) {
+    stop("Please provide a valid location name as a string.")
+  }
+  
+  # Validate geocoding method
+  supported_methods <- c("osm", "census", "arcgis", "census_simple", "geocodio",
+                         "mapbox", "google", "bing", "here", "tomtom", "nominatim", "tiger")
+  if (!(method %in% supported_methods)) {
```
Comment (koenderks) [inline R/geocode.R]: You can use `match.arg(method)` to automatically show this error, see https://www.r-at-work.com/functions/base/match-arg.html.

File: R/geocode.R [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    
+    # Create a city-like df with geocoded coordinates
+    split_locations <- strsplit(coords$display_name, split = ",")[[1]]
+    
+    city <- data.frame(
+      name = split_locations[1],
+      country = split_locations[2],
+      lat = coords$lat,
+      long = coords$lon,
+      population = NA
```
Comment (koenderks) [inline R/geocode.R]: I don't think you need to make a column for `population`, it is mostly used by the twitter bot to filter out small cities.

File: R/geocode.R [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+#' \dontrun{
+#'   # Geocode a location using OpenStreetMap (default)
+#'   geocode_raw("New York, USA")
+#'
+#'   # Geocode a location using Google (API key required)
+#'   geocode_raw("Tokyo, Japan", method = "google")
+#' }
+#'
+#' @export
+geocode <- function(location, method = "osm") {
```
Comment (koenderks) [inline R/geocode.R]: make a hidden function `.geocode` for consistency.

**Review by koenderks — CHANGES_REQUESTED** [review]
Nice, I wonder where this should be implemented in the existing workflow. It would be nice if the geocoding is done automatically if the city name is not found in the internal database. I guess you can try the geocode at [this line](https://github.com/koenderks/rcityviews/blob/7578d7028c32beeda68ff24ba9a97d72a517ec92/R/helpers.R#L51) and then adjust [these lines](https://github.com/koenderks/rcityviews/blob/7578d7028c32beeda68ff24ba9a97d72a517ec92/R/helpers.R#L31-L33) a bit so that it handles the output of the geocode and also fails gracefully. Anyway, I made some comments that may help to reduce dependencies and enhance readability.

File: R/geocode.R (hunk 1 above)
Comment (pokyah) [reply R/geocode.R]: I could use the `.geocode `function as a fallback if no location found in the database but then how to let the user choose the geocoding service ? Should me prompt him/her and also provide an explanation in the README about the API keys ? Or do we systematically use the default free `osm `geocoder ?

File: R/geocode.R (hunk 4 above)
Comment (pokyah) [reply R/geocode.R]: Also I wonder if `new_city` will still be useful if I use .geocode as a fallback. What do you think?

Comment (koenderks) [reply R/geocode.R]: You could also use the geocode as a fallback for not using coordinates `lat` and `long` in the `new_city` function?

So, attempt geocoding in this case:

```
city <- new_city(name = "Namur", country = "Belgium")
```

but not in this case:

```
city <- new_city(name = "Namur", country = "Belgium", lat = 50.4669, long = 4.86746)
```

You could then either add the `method` argument to the `new_city` function or use something like `options(rcityviews.geocode.method)`. That way, we move all functionality regarding unknown cities that are not in the database to the `new_city()` function.

File: R/geocode.R (hunk 2 above)
Comment (pokyah) [reply R/geocode.R]: implemented in my last commit

File: R/geocode.R (hunk 3 above)
Comment (pokyah) [reply R/geocode.R]: removed in my last commit

File: R/geocode.R (hunk 4 above)
Comment (pokyah) [reply R/geocode.R]: OK. Thanks for your feedback. Will implement as it !

Comment (pokyah) [reply R/geocode.R]: if the city does not exists in the database, I guess that the best workflow would be : 

- to force the user to use `new_city` with implemented geocoder rather than implement an automagic fallback.

**Comment by pokyah** [comment]
Example of the current implementation  : 

```
map = cityview(
 name = "Ixelles",
 license = F,
 theme = "vintage",
timeout = 1000,
legend = F,
zoom = 3.5,
border = "circle",
cache = T)
```
returns (as expected): 

```
Error in .resolveConflicts(name, indexes, dataset) : 
  There is no city called 'Ixelles' in the available data.
 Use new_city function() to geocode the location
```

thus using new_city with implemented geocoder : 

```
name = new_city(name = "Ixelles", country = "Belgium")
map = cityview(
 name = name ,
 license = F,
 theme = "vintage",
timeout = 1000,
legend = F,
zoom = 3.5,
border = "circle",
cache = T)
```
perfectly works : 

```
[=======================================] 1/1 (100%) Elapsed:  1s Remaining:  0s
Discovered the city of Ixelles, Belgium at 50.8331141° / 4.3668279°!

[=======================================] 1/1 (100%) Elapsed:  1s Remaining:  0s
Discovered the city of Ixelles, Belgium at 50.8331141° / 4.3668279°!
```

**Comment by koenderks** [comment]
Can you perform a rebase on the `development` branch to get the most up-to-date code?

Open terminal:

```
git fetch upstream
git rebase upstream/developent
git push
```

If you encounter any conflicts with the `daily.png` and `name.txt` file, simply choose the `use the modified file` option.

File: R/new_city.R [hunk 5]
```diff
@@ -48,9 +51,16 @@
...
-new_city <- function(name = NULL, country = NULL, lat = NULL, long = NULL) {
-  stopifnot("specify all input arguments" = all(c(!is.null(name), !is.null(country), !is.null(lat), !is.null(long))))
-  stopifnot("all input must be of length 1" = all(c(length(name) == 1, length(country) == 1, length(lat) == 1, length(long) == 1)))
+new_city <- function(name = NULL, country = NULL, lat = NULL, long = NULL, method = "osm") {
+  # Use match.arg to validate the method parameter
+  method <- match.arg(method, choices = c("osm", "census", "arcgis", "census_simple", "geocodio",
+                                          "mapbox", "google", "bing", "here", "tomtom", "nominatim", "tiger"))
+  stopifnot("At least provide a location name and country" = all(c(length(name) == 1, length(country) == 1)))
+  if(is.null(lat) || is.null(long)){
+    out = .geocode(name, country)
```
Comment (koenderks) [inline R/new_city.R]: Please use `<-` instead of `=`.

Comment (koenderks) [reply R/new_city.R]: Also run `styler::style_pkg()` in the R console when your working directory is set to the package folder to automatically adhere to the tidyverse style guide (https://style.tidyverse.org)

File: R/helpers.R [hunk 6]
```diff
@@ -13,6 +13,46 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+.geocode <- function(name, country, method = "osm") {
+  if (missing(name) || !is.character(name)) {
+    stop("Please provide a valid location name as a string.")
+  }
+  
+  # Use match.arg to validate the method parameter
+  method <- match.arg(method, choices = c("osm", "census", "arcgis", "census_simple", "geocodio",
```
Comment (koenderks) [inline R/helpers.R]: You don't need this call to `match.arg()` if you do this in the top-level `new_city()` function, but then you need to pass `.geocode(name, country, method)` in the `new_city()` function.

File: R/helpers.R [hunk 7]
```diff
@@ -13,6 +13,46 @@
...
+  if (method %in% methods_with_keys) {
+    api_key_env <- paste0(toupper(method), "_API_KEY")
+    api_key <- Sys.getenv(api_key_env)
+    if (api_key == "") {
+      stop(paste0("API key for ", method, " is required. Please set the '", api_key_env, "' environment variable."))
+    }
+  }
+  
+  result <- tryCatch({
+    geocode_df <- tibble::tibble(address = paste0(name, " ", country)) %>%
```
Comment (koenderks) [inline R/helpers.R]: Can you still get rid of these dependencies via `tidygeocoder::geocode(.tbl = data.frame(address=location), "address", method = method, quiet = TRUE)`, you probably also need `address <- NULL` at the top of the function to silence the `geocode: no visible binding for global variable ‘address’` note in https://github.com/koenderks/rcityviews/actions/runs/11346897862/job/31557368996#step:9:117

File: R/cityview.R
Comment (koenderks) [inline R/cityview.R:1]: The changes in this file can be reverted, right? Or is there anything different from before?

File: R/new_city.R [hunk 8]
```diff
@@ -48,9 +51,16 @@
 #' }
 #' @export
 
-new_city <- function(name = NULL, country = NULL, lat = NULL, long = NULL) {
-  stopifnot("specify all input arguments" = all(c(!is.null(name), !is.null(country), !is.null(lat), !is.null(long))))
-  stopifnot("all input must be of length 1" = all(c(length(name) == 1, length(country) == 1, length(lat) == 1, length(long) == 1)))
+new_city <- function(name = NULL, country = NULL, lat = NULL, long = NULL, method = "osm") {
```
Comment (koenderks) [inline R/new_city.R]: `method = c("osm", "census", "arcgis", "census_simple", "geocodio",
                                           "mapbox", "google", "bing", "here", "tomtom", "nominatim", "tiger")`

File: R/new_city.R [hunk 9]
```diff
@@ -48,9 +51,16 @@
...
+  # Use match.arg to validate the method parameter
+  method <- match.arg(method, choices = c("osm", "census", "arcgis", "census_simple", "geocodio",
```
Comment (koenderks) [inline R/new_city.R]: `method <- match.arg(method)`

File: R/helpers.R [hunk 10]
```diff
@@ -13,6 +13,46 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+.geocode <- function(name, country, method = "osm") {
+  if (missing(name) || !is.character(name)) {
+    stop("Please provide a valid location name as a string.")
```
Comment (koenderks) [inline R/helpers.R]: `stopifnot("Please provide a location name as a string" = is.character(name))`
Since name cannot be empty

File: R/helpers.R [hunk 11]
```diff
@@ -48,15 +91,12 @@
 .resolveConflicts <- function(name, indexes, dataset) {
   index <- indexes
   if (length(indexes) == 0) {
-    stop(paste0("There is no city called '", name, "' in the available data.\nUse 'new_city()' or create an issue including lat/long coordinates at https://github.com/koenderks/rcityviews/issues."))
+    stop(paste0("There is no city called '", name, "' in the available data.\n Use new_city function() to geocode the location"))
```
Comment (koenderks) [inline R/helpers.R]: `Use the 'new_city()' function to...`

File: man/new_city.Rd [hunk 12]
```diff
@@ -14,6 +14,10 @@ new_city(name = NULL, country = NULL, lat = NULL, long = NULL)
 \item{lat}{A single numeric value to be used as the latitude.}
 
 \item{long}{A single numeric value to be used as the longitude.}
+
+\item{method}{A character string specifying the geocoding method to use. Defaults to "osm" (OpenStreetMap). Supported methods include:
```
Comment (koenderks) [inline man/new_city.Rd]: `...to use when the user does not specify \code{lat} or \code{long}.`

**Review by koenderks — CHANGES_REQUESTED** [review]

File: R/new_city.R (hunk 5 above)
Comment (pokyah) [reply R/new_city.R]: Done

File: R/helpers.R (hunk 6 above)
Comment (pokyah) [reply R/helpers.R]: you are totally right !

File: R/helpers.R (hunk 7 above)
Comment (pokyah) [reply R/helpers.R]: Sure !

File: R/new_city.R (hunk 8 above)
Comment (pokyah) [reply R/new_city.R]: What do you want here?

File: man/new_city.Rd (hunk 12 above)
Comment (pokyah) [reply man/new_city.Rd]: Yes better!

File: R/helpers.R (hunk 11 above)
Comment (pokyah) [reply R/helpers.R]: Clearer

File: R/helpers.R (hunk 10 above)
Comment (pokyah) [reply R/helpers.R]: cleaner indeed!

File: R/cityview.R
Comment (pokyah) [reply R/cityview.R:1]: Which changes are you talking about?

File: R/new_city.R (hunk 8 above)
Comment (pokyah) [reply R/new_city.R]: ok linked to following point

File: R/new_city.R (hunk 9 above)
Comment (pokyah) [reply R/new_city.R]: done

File: R/helpers.R [hunk 13]
```diff
@@ -13,6 +13,65 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+.geocode <- function(name, country, method = "osm") {
```
Comment (koenderks) [inline R/helpers.R]:
```suggestion
.geocode <- function(name, country, method) {
```

File: R/helpers.R [hunk 14]
```diff
@@ -13,6 +13,65 @@
...
+        long = geocode_df$long[1]
+      )
+    },
+    error = function(e) {
+      stop("Geocoding error: ", e$message)
+    }
+  )
+
+  return(result)
+}
```
Comment (koenderks) [inline R/helpers.R]:
```suggestion
```

Comment (koenderks) [reply R/helpers.R]: This tryCatch statement is now twice in the code

File: R/new_city.R [hunk 15]
```diff
@@ -49,9 +52,16 @@
 #' }
 #' @export
 
-new_city <- function(name = NULL, country = NULL, lat = NULL, long = NULL) {
-  stopifnot("specify all input arguments" = all(c(!is.null(name), !is.null(country), !is.null(lat), !is.null(long))))
-  stopifnot("all input must be of length 1" = all(c(length(name) == 1, length(country) == 1, length(lat) == 1, length(long) == 1)))
+new_city <- function(name = NULL, country = NULL, lat = NULL, long = NULL, method = "osm") {
```
Comment (koenderks) [inline R/new_city.R]:
```suggestion
new_city <- function(name = NULL, country = NULL, lat = NULL, long = NULL, method = c("osm", "census", "arcgis", "census_simple", "geocodio", "mapbox", "google", "bing", "here", "tomtom", "nominatim", "tiger")) {
```

File: R/new_city.R [hunk 16]
```diff
@@ -49,9 +52,16 @@
...
+  # Use match.arg to validate the method parameter
+  method <- c("osm", "census", "arcgis", "census_simple", "geocodio", "mapbox", "google", "bing", "here", "tomtom", "nominatim", "tiger")
```
Comment (koenderks) [inline R/new_city.R]:
```suggestion
```

File: png/name.txt [hunk 17]
```diff
@@ -1 +1,8 @@
-Goiania, Brazil
\ No newline at end of file
+<<<<<<< HEAD
+Ploiesti, Romania
+=======
+Barddhaman, India
+
+=======
+Barddhaman, India
+>>>>>>> 5a391a1dd4790303777d0e7e39c3d720ba80d891
```
Comment (koenderks) [inline png/name.txt]:
```suggestion
Goiania, Brazil
```

**Review by koenderks — CHANGES_REQUESTED** [review]

File: R/helpers.R (hunk 14 above)
Comment (pokyah) [reply R/helpers.R]: Sorry for that!

File: R/new_city.R (hunk 16 above)
Comment (pokyah) [reply R/new_city.R]: done

File: R/helpers.R [hunk 18]
```diff
@@ -13,6 +13,43 @@
...
+
 .getCity <- function(name) {
```
Comment (koenderks) [inline R/helpers.R:54]: I thought the idea was that the geocoding was only used in the `new_city()` function and we force the user to use that function if the city is not in the internal database? Anyway, I would prefer it that way since the changes are then minimal. Then you only have to adjust the stop message like you did, and can revert the other changes.

File: R/new_city.R
Comment (koenderks) [inline R/new_city.R:1]: You have to adjust the `@usage` line to reflect the addition of the `method` argument, then run `roxygen2::roxygenize()` to silence the following warning:

```
❯ checking for code/documentation mismatches ... WARNING
  Codoc mismatches from Rd file 'new_city.Rd':
  new_city
    Code: function(name = NULL, country = NULL, lat = NULL, long = NULL,
                   method = c("osm", "census", "arcgis", "census_simple",
                   "geocodio", "mapbox", "google", "bing", "here",
                   "tomtom", "nominatim", "tiger"))
    Docs: function(name = NULL, country = NULL, lat = NULL, long = NULL)
    Argument names in code not in docs:
      method

❯ checking Rd \usage sections ... WARNING
  Documented arguments not in \usage in Rd file 'new_city.Rd':
    ‘method’
  
  Functions with \usage entries need to have the appropriate \alias
  entries, and all their arguments documented.
  The \usage entries must correspond to syntactically valid R code.
  See chapter ‘Writing R documentation files’ in the ‘Writing R
  Extensions’ manual.
```

You can check this using `devtools::check(args = "--no-multiarch", cran = TRUE)`

File: R/helpers.R [hunk 19]
```diff
@@ -13,6 +13,43 @@
...
+  }
+
+  result <- tryCatch(
+    {
+      geocode_df <- tidygeocoder::geocode(
+        data.frame(address = paste0(name, " ", country)), "address",
+        method = method, quiet = TRUE
+      )
+      if (any(is.na(geocode_df$lat)) || any(is.na(geocode_df$long))) {
+        stop("Geocoding failed: Unable to find coordinates for the provided location name.")
```
Comment (koenderks) [inline R/helpers.R]: Maybe add something like `..., manually enter latitude and longitude coordinates using the 'lat' and 'long' arguments`

File: R/helpers.R [hunk 20]
```diff
@@ -13,6 +13,43 @@
...
+      stop(paste0("API key for ", method, " is required. Please set the '", api_key_env, "' environment variable."))
+    }
+  }
+
+  result <- tryCatch(
+    {
+      geocode_df <- tidygeocoder::geocode(
+        data.frame(address = paste0(name, " ", country)), "address",
+        method = method, quiet = TRUE
+      )
```
Comment (koenderks) [inline R/helpers.R]:
```suggestion
      geocode_df <- tidygeocoder::geo(
        city = name, country = country, method = method, quiet = TRUE, progress_bar = FALSE, limit = 1
      )
```

Comment (koenderks) [reply R/helpers.R]: Uses lower-level function `geo()` with less data wrangling

File: R/helpers.R [hunk 21]
```diff
@@ -30,9 +67,13 @@
       indexes <- which(dataset[["name"]] == name)
       index <- .resolveConflicts(name, indexes, dataset)
       if (is.null(index)) {
-        return(NULL)
+        city <- .geocode(
+          name = name$name, country = name$country
+        )
+      } else {
+        city <- dataset[index, ]
```
Comment (koenderks) [inline R/helpers.R]:
```suggestion
        return(NULL)
```

File: R/helpers.R [hunk 22]
```diff
@@ -30,9 +67,13 @@
...
       }
-      city <- dataset[index, ]
+      return(city)
```
Comment (koenderks) [inline R/helpers.R]:
```suggestion
      city <- dataset[index, ]
```

File: R/helpers.R [hunk 23]
```diff
@@ -50,15 +91,12 @@
...
   index <- indexes
   if (length(indexes) == 0) {
-    stop(paste0("There is no city called '", name, "' in the available data.\nUse 'new_city()' or create an issue including lat/long coordinates at https://github.com/koenderks/rcityviews/issues."))
+    stop(paste0("There is no city called '", name, "' in the available data.\n Use the 'new_city()' function to geocode the location"))
   } else if (length(indexes) > 1) {
     selection <- utils::menu(
       choices = paste0(dataset[indexes, 1], ", ", dataset[indexes, 2], " | Lat: ", round(dataset[indexes, 3], 3), " | Long: ", round(dataset[indexes, 4], 3)),
       title = "More than one city matched to this name, which one to pick?"
     )
-    if (selection == 0) {
```
Comment (koenderks) [inline R/helpers.R:59]: Revert these changes

**Review by koenderks — CHANGES_REQUESTED** [review]

File: R/helpers.R [hunk 24]
```diff
@@ -13,6 +13,43 @@
...
+.geocode <- function(name, country, method) {
+  stopifnot("Please provide a location name as a string" = is.character(name))
+
+  # Handle API keys for methods that require them
+  methods_with_keys <- c("google", "bing", "here", "tomtom", "mapbox", "geocodio")
+  if (method %in% methods_with_keys) {
+    api_key_env <- paste0(toupper(method), "_API_KEY")
+    api_key <- Sys.getenv(api_key_env)
+    if (api_key == "") {
+      stop(paste0("API key for ", method, " is required. Please set the '", api_key_env, "' environment variable."))
```
Comment (koenderks) [inline R/helpers.R]:
```suggestion
      stop(paste0("API key for ", method, " is required. Please set the '", api_key_env, "' environment variable using usethis::edit_r_environ() to open your .Renviron file and add the API key."))
```

File: R/helpers.R (hunk 19 above)
Comment (pokyah) [reply R/helpers.R]: Good idea!

File: R/helpers.R (hunk 20 above)
Comment (pokyah) [reply R/helpers.R]: I prefer to use the following: 
```

      geocode_df <- tidygeocoder::geo(
        address = paste0(name, " ", country),
        method = method, quiet = TRUE, progress_bar = FALSE, limit = 1
      )
```
Which allow to pass any location rather than being restricted to city names. Doing so allows to geocode a neighborhood name for example.

Want do you think?

Comment (koenderks) [reply R/helpers.R]: > I prefer to use the following:
> 
> ```
> 
>       geocode_df <- tidygeocoder::geo(
>         address = paste0(name, " ", country),
>         method = method, quiet = TRUE, progress_bar = FALSE, limit = 1
>       )
> ```
> 
> Which allow to pass any location rather than being restricted to city names. Doing so allows to geocode a neighborhood name for example.
> 
> Want do you think?

Good point!

File: R/helpers.R (hunk 24 above)
Comment (pokyah) [reply R/helpers.R]: Better!

File: R/helpers.R (hunk 18 above)
Comment (pokyah) [reply R/helpers.R:54]: I'm not sure we changes you want to revert. I would prefer to not use the geocoding fallback `.getCity` function and would therefore restore .getCity function as it was designed originally : 

```
.getCity <- function(name) {
  if (is.null(name)) {
    city <- .randomCity(NULL)
  } else {
    if (inherits(name, "rcityviewsCity")) {
      city <- name
    } else if (inherits(name, "data.frame")) {
      stopifnot("input data frame is missing 'name' column" = "name" %in% colnames(name))
      stopifnot("input data frame is missing 'country' column" = "country" %in% colnames(name))
      stopifnot("input data frame is missing 'lat' column" = "lat" %in% colnames(name))
      stopifnot("input data frame is missing 'long' column" = "long" %in% colnames(name))
      city <- name
    } else {
      dataset <- rcityviews::cities
      indexes <- which(dataset[["name"]] == name)
      index <- .resolveConflicts(name, indexes, dataset)
      if (is.null(index)) {
        return(NULL)
      }
      city <- dataset[index, ]
    }
  }
  return(city)
}
```

File: R/helpers.R (hunk 21 above)
Comment (pokyah) [reply R/helpers.R]: Have restore .getCity to its original state without using the fallback geocoding.

File: R/new_city.R
Comment (pokyah) [reply R/new_city.R:1]: Here is the output : 


```
> devtools::check(args = "--no-multiarch", cran = TRUE)
══ Documenting ══════════════════════════════════════════════════════════
ℹ Updating rcityviews documentation
ℹ Loading rcityviews
Data by © OpenStreetMap contributors
Writing new_city.Rd

══ Building ═════════════════════════════════════════════════════════════
Setting env vars:
• CFLAGS    : -Wall -pedantic
• CXXFLAGS  : -Wall -pedantic
• CXX11FLAGS: -Wall -pedantic
• CXX14FLAGS: -Wall -pedantic
• CXX17FLAGS: -Wall -pedantic
• CXX20FLAGS: -Wall -pedantic
── R CMD build ──────────────────────────────────────────────────────────
✔  checking for file 'C:\Users\pokyah\Projects\rcityviews/DESCRIPTION' (469ms)
─  preparing 'rcityviews': (1.6s)
✔  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts (551ms)
─  checking for empty or unneeded directories
   Removed empty directory 'rcityviews/vignettes'
─  building 'rcityviews_1.1.2.tar.gz'
   
══ Checking ═════════════════════════════════════════════════════════════
Setting env vars:
• _R_CHECK_CRAN_INCOMING_REMOTE_               : FALSE
• _R_CHECK_CRAN_INCOMING_                      : FALSE
• _R_CHECK_FORCE_SUGGESTS_                     : FALSE
• _R_CHECK_PACKAGES_USED_IGNORE_UNUSED_IMPORTS_: FALSE
• NOT_CRAN                                     : true
── R CMD check ──────────────────────────────────────────────────────────
─  using log directory 'C:/Users/pokyah/AppData/Local/Temp/RtmpmCC111/file444c66bc3a61/rcityviews.Rcheck' (561ms)
─  using R version 4.3.3 (2024-02-29 ucrt)
─  using platform: x86_64-w64-mingw32 (64-bit)
─  R was compiled by
       gcc.exe (GCC) 12.3.0
       GNU Fortran (GCC) 12.3.0
─  running under: Windows 11 x64 (build 22631)
─  using session charset: UTF-8
─  using options '--no-manual --as-cran'
✔  checking for file 'rcityviews/DESCRIPTION'
─  this is package 'rcityviews' version '1.1.2'
─  package encoding: UTF-8
✔  checking package namespace information
✔  checking package dependencies (1.6s)
✔  checking if this is a source package ...
✔  checking if there is a namespace
✔  checking for .dll and .exe files
✔  checking for hidden files and directories
✔  checking for portable file names
─  checking whether package 'rcityviews' can be installed ... [35s] OK (35s)
✔  checking package directory
✔  checking for future file timestamps ... 
✔  checking DESCRIPTION meta-information (347ms)
✔  checking top-level files
✔  checking for left-over files ...
✔  checking index information
✔  checking package subdirectories (359ms)
✔  checking R files for non-ASCII characters ... 
✔  checking R files for syntax errors ... 
─  checking whether the package can be loaded ... [11s] OK (11.5s)
─  checking whether the package can be loaded with stated dependencies ... [11s] OK (11.1s)
─  checking whether the package can be unloaded cleanly ... [13s] OK (12.9s)
─  checking whether the namespace can be loaded with stated dependencies ... [13s] OK (13.1s)
─  checking whether the namespace can be unloaded cleanly ... [12s] OK (12.2s)
─  checking loading without being on the library search path ... [11s] OK (11.1s)
─  checking startup messages can be suppressed ... [11s] OK (23.5s)
N  checking dependencies in R code (11s)
   There are ::: calls to the package's namespace in its code. A package
     almost never needs to use ::: for its own objects:
     '.buildCity' '.getBoundaries' '.nonMemoiseRequestData' '.randomCity'
     '.themeOptions'
✔  checking S3 generic/method consistency (9.4s)
✔  checking replacement functions (9.9s)
✔  checking foreign function calls (10.7s)
─  checking R code for possible problems ... [51s] OK (50.7s)
✔  checking Rd files ... 
✔  checking Rd metadata ... 
✔  checking Rd line widths ... 
✔  checking Rd cross-references ... 
✔  checking for missing documentation entries (10.7s)
✔  checking for code/documentation mismatches (29.1s)
✔  checking Rd \usage sections (12.7s)
✔  checking Rd contents ... 
✔  checking for unstated dependencies in examples ... 
✔  checking contents of 'data' directory ... 
✔  checking data for non-ASCII characters (343ms)
✔  checking LazyData
✔  checking data for ASCII and uncompressed saves ... 
─  checking examples ... [11s] OK (11.4s)
✔  checking for unstated dependencies in 'tests' ... 
─  checking tests ...
    [10s] OKtestthat.R'
   * checking for non-standard things in the check directory ... OK
   * checking for detritus in the temp directory ... OK
   * DONE
   
   Status: 1 NOTE
   See
     'C:/Users/pokyah/AppData/Local/Temp/RtmpmCC111/file444c66bc3a61/rcityviews.Rcheck/00check.log'
   for details.
   
   
── R CMD check results ──────────────────────────── rcityviews 1.1.2 ────
Duration: 5m 4.7s

❯ checking dependencies in R code ... NOTE
  There are ::: calls to the package's namespace in its code. A package
    almost never needs to use ::: for its own objects:
    '.buildCity' '.getBoundaries' '.nonMemoiseRequestData' '.randomCity'
    '.themeOptions'

0 errors ✔ | 0 warnings ✔ | 1 note ✖
```

Comment (pokyah) [reply R/new_city.R:1]: So I should update the @usage element accordingly

File: R/helpers.R (hunk 23 above)
Comment (pokyah) [reply R/helpers.R:59]: I prefer to work with `new_city` rather than any fallback

**Review by koenderks — CHANGES_REQUESTED** [review]
Ok I think we are almost done. Can you please make a pull request to the new branch `geocoding`? Probably best to take this branch and then replace the `DESCRIPTION`, `helpers.R`, and `new_city.R` files in it. Because the `development` branch updates every 3 hours, there will most of the time be merge conflicts so this is probably the easiest way.

**Review by koenderks — APPROVED** [review]

**Comment by koenderks** [comment]
Merging this into `geocoding` and then into `development`, thanks! For the themes, I prefer to keep only the 10 rcityviews themes (prettymaps is nice, but I want to keep rcityviews something original), but I have made a persistent storage mechanism using the `city_themes()` function if you want to store them!

**Comment by pokyah** [comment]
Not sure what you want me to do with the new geocoding branch?
Can you explain me the full workflow?

**Comment by koenderks** [comment]
Nothing ;) I diverted your pull request to https://github.com/koenderks/rcityviews/tree/geocoding and merged it into that branch! I just merged the geocoding branch into development, so you should have the new functionality in the package if you run.

```
remotes::install_github("koenderks/rcityviews", dependencies = TRUE)
```

Thanks for the effort and the new addition, it is very cool!

**Comment by pokyah** [comment]
> Nothing ;) I diverted your pull request to https://github.com/koenderks/rcityviews/tree/geocoding and merged it into that branch! I just merged the geocoding branch into development, so you should have the new functionality in the package if you run.
> 
> ```
> remotes::install_github("koenderks/rcityviews", dependencies = TRUE)
> ```
> 
> Thanks for the effort and the new addition, it is very cool!

Thanks for letting me contribute to the package! I really enjoyed contributing on in and have learnt quite a few things on the way!
